src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ConnectionPool/ConnectionSetupDiagnostics.cs

```diff
@@ -0,0 +1,98 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Threading;
+
+namespace System.Net.Http
+{
+    // Implements distributed tracing logic for managing the "HTTP connection_setup" and "HTTP wait_for_connection" Activities.
+    internal static class ConnectionSetupDiagnostics
+    {
+        private static readonly ActivitySource s_connectionsActivitySource = new ActivitySource(DiagnosticsHandlerLoggingStrings.ConnectionsNamespace);
+
+        public static Activity? StartConnectionSetupActivity(bool isSecure, HttpAuthority authority)
+        {
+            Activity? activity = null;
+            if (s_connectionsActivitySource.HasListeners())
+            {
+                // Connection activities should be new roots and not parented under whatever
+                // request happens to be in progress when the connection is started.
+                Activity.Current = null;
+                activity = s_connectionsActivitySource.StartActivity(DiagnosticsHandlerLoggingStrings.ConnectionSetupActivityName);
+            }
+
+            if (activity is not null)
+            {
+                activity.DisplayName = $"HTTP connection_setup {authority.HostValue}:{authority.Port}";
+                if (activity.IsAllDataRequested)
+                {
+                    activity.SetTag("server.address", authority.HostValue);
+                    activity.SetTag("server.port", authority.Port);
+                    activity.SetTag("url.scheme", isSecure ? "https" : "http");
+                }
+            }
+
+            return activity;
+        }
+
+        public static void StopConnectionSetupActivity(Activity activity, Exception? exception, IPEndPoint? remoteEndPoint)
+        {
+            Debug.Assert(activity is not null);
```

Comment:
See previous comment. IMO if we want to get rid of this pattern because we prefer simplicity/readability over saving a few cycles, we should do it for all telemetry not only distributed tracing.

Reply:
cc @stephentoub

Reply:
Actually, I'm being inconsistent since the Start() methods also would need the check here. I will wait for more feedback before changing anything about this.

Reply:
`ReportError` has the check inside and not at the call site, that's one that is bothering me (or rather Stop and AddConnectionLink). With Start, the check and code is more complicated so it makes sense to keep it inside.

Reply:
`ReportError` is already on an expensive Exception path, so no point bothering with micro-optimizations.
I leave this for now as-is, but I believe we should revisit our telemetry coding patterns/guidelines.

My preference would be clarity over microoptimizations everywhere, but some may disagree.